autoscale/src/main/java/com/sequenceiq/periscope/service/MandatoryScalingAdjustmentService.java

```diff
@@ -0,0 +1,169 @@
+import com.sequenceiq.periscope.domain.LoadAlertConfiguration;
+import com.sequenceiq.periscope.model.MandatoryScalingAdjustmentParameters;
+import com.sequenceiq.periscope.model.yarn.YarnScalingServiceV1Response;
+import com.sequenceiq.periscope.monitor.client.YarnMetricsClient;
+import com.sequenceiq.periscope.monitor.evaluator.load.YarnResponseUtils;
+import com.sequenceiq.periscope.monitor.sender.ScalingEventSender;
+import com.sequenceiq.periscope.utils.ClusterUtils;
+import com.sequenceiq.periscope.utils.StackResponseUtils;
+import com.sequenceiq.periscope.utils.TimeUtil;
+
+@Service
+public class MandatoryScalingAdjustmentService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MandatoryScalingAdjustmentService.class);
+
+    @Inject
+    private ScalingEventSender scalingEventSender;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private StackResponseUtils stackResponseUtils;
+
+    @Inject
+    private YarnMetricsClient yarnMetricsClient;
+
+    @Inject
+    private YarnResponseUtils yarnResponseUtils;
+
+    private Cluster cluster;
+
+    private LoadAlert loadAlert;
+
+    private LoadAlertConfiguration loadAlertConfiguration;
+
+    private String policyHostGroup;
+
+    public void performMandatoryAdjustment(Long clusterId, String pollingUserCrn, StackV4Response stackResponse,
+            MandatoryScalingAdjustmentParameters scalingAdjustmentParameters) {
+
+        cluster = clusterService.findById(clusterId);
+        loadAlert = cluster.getLoadAlerts().iterator().next();
+        loadAlertConfiguration = loadAlert.getLoadAlertConfiguration();
+        policyHostGroup = loadAlert.getScalingPolicy().getHostGroup();
+
+        Map<String, String> hostFqdnsToInstanceId = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackResponse, policyHostGroup);
+        List<String> servicesHealthyHostInstanceIds = stackResponseUtils.getCloudInstanceIdsWithServicesHealthyForHostGroup(stackResponse,
+                policyHostGroup);
+        List<String> stoppedHostIds = stackResponseUtils.getStoppedCloudInstanceIdsInHostGroup(stackResponse, policyHostGroup);
+
+        int stoppedNodeCount = stoppedHostIds.size();
+        int existingClusterNodeCount = stackResponse.getNodeCount() - stoppedNodeCount;
+
+        publishScalingEventIfNeeded(hostFqdnsToInstanceId, servicesHealthyHostInstanceIds, existingClusterNodeCount, stoppedHostIds, pollingUserCrn,
+                stackResponse, scalingAdjustmentParameters);
+    }
+
+    private void publishScalingEventIfNeeded(Map<String, String> hostFqdnsToInstanceId, List<String> hostIdsWithServicesHealthy, int existingClusterNodeCount,
+            List<String> stoppedHostIds, String pollingUserCrn, StackV4Response stackV4Response,
+            MandatoryScalingAdjustmentParameters scalingAdjustmentParameters) {
+
+        if (upscalable(scalingAdjustmentParameters.getMandatoryStopStartUpscaleAdjustment())) {
```

Comment:
General question: In the case of mandatory operations - should we ignore the cooldown parameters ? We're trying to get the cluster into a desired state as a response to user input. cooldown should not matter here.
(The code separation makes this so much easier to reason about)

Reply:
I had respected coolDownMinutes in these adjustments to be equivalent to how scalingRequests were issued previously, and since `coolDownMinutes` pertains to the duration between any kind of scalingActivity. Updated to ignore the coolDownParams in case of manualAdjustments as suggested.